add torch.diff support for n greater than 1

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67260

Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

mikaylagawarecki committed:
add diff support for n greater than 1

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] add diff support for n greater than 1"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] add diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] add diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
-  if (self.dtype() == at::kBool) {
-    return at::logical_xor(at::narrow(self, dim, 1, out_len), at::narrow(self, dim, 0, out_len));
+  auto result = self;
+  bool is_kBool = self.dtype() == at::kBool;
```
nit: `(self.dtype() == at::kBool;)` for readability

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
 }
 
 static inline Tensor diff_helper(const Tensor& self, int64_t n, int64_t dim) {
+  if (n == 0){
```
Comment here

Code formatting nit: `(n == 0) {`

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
+  for (const auto i : c10::irange(n)) {
+    (void)i; //Suppress unused variable warning
+    if (is_kBool){
+      result = at::logical_xor(at::narrow(result, dim, 1, out_len), at::narrow(result, dim, 0, out_len));
```
Nice bool handling

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
-  auto out_len = self.size(dim) - 1;
+  if (n == 0){
+    at::native::resize_output(result, self.sizes());
+    check_scalar_type_device_layout_equal(result, self);
```
This check is a little severe but that seems OK

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
+  }
+
+  n = n >= self.size(dim) ? self.size(dim) : n;
+  auto out_len = self.size(dim) - n;
```
nit: `const auto`

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
+
   if (self.dtype() == at::kBool) {
-    return at::logical_xor_out(result, at::narrow(self, dim, 1, out_len), at::narrow(self, dim, 0, out_len));
+    at::logical_xor_out(result, at::narrow(prev_result, dim, 1, out_len), at::narrow(prev_result, dim, 0, out_len));
```
Nice pattern of using `_out` variants for the final computation

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
         ((S, S), 0, None, (2, S)),
         ((S, S, S), 1, None, None),
-        ((S, S, S), 1, (S, 1, S), (S, 1, S)),)
+        ((S, S, S), 2, None, None),
```
Great extension

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
+        prepend_size = 0 if (size_prepend is None) else size_prepend[dim]
+        append_size = 0 if (size_append is None) else size_append[dim]
+        dim_size = size[dim] + prepend_size + append_size
+        for n in range(dim_size + offset):  # test that there are no erors when n > dim_size
```
This is very thorough (which is great!) but also computationally expensive (which is less great). How about including `n` in the sample input for each sample and just including 1-2 samples where `n > dim_size`?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
+        append_size = 0 if (size_append is None) else size_append[dim]
+        dim_size = size[dim] + prepend_size + append_size
+        for n in range(dim_size + offset):  # test that there are no erors when n > dim_size
+            args = (make_tensor(size, device, dtype,
```
nit: sometimes it's faster to partial the function like:

```
make_arg = partial(make_tensor, dict(device=device, dtype=dtype, requires_grad=requires_grad))
```

then these lines would be just 

```
make_arg(size_prepend) if size_prepend else None
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
+        for n in range(dim_size + offset):  # test that there are no erors when n > dim_size
+            args = (make_tensor(size, device, dtype,
+                                low=None, high=None,
+                                requires_grad=requires_grad), n, dim,
```
Hiding `n` and `dim` back here is tough to read. Passing them as kwargs would be more readable:

```
s = SampleInput(make_arg(size), kwargs={
  "n": n,
  "dim": dim,
  ...   
})
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7920,6 +7928,11 @@ def ref_pairwise_distance(input1, input2):
                    safe_casts_outputs=True),
     OpInfo('diff',
            op=torch.diff,
+           # np.diff has np._NoValue as default values for prepend and append, compare_with_reference breaks if prepend/append
```
Really nice reference handling

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all -l ciflow/slow

mruberry reviewed (approved):
Hey @mikaylagawarecki! This looks great. I left a few comments/suggestions for your review. Mostly to consider slimming the sample inputs so the tests don't take so long. 

I also added some additional tests, so let's wait for those to complete.

Thanks for patiently waiting for the review! Moving was a lot more time consuming than I expected. As soon as the tests pass this should be OK to land; ping me if you have any additional questions.

mikaylagawarecki committed:
Update on "add torch.diff support for n greater than 1"


Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31930294).
